pkg/oci/routes.go

```diff
@@ -0,0 +1,291 @@
+			NetworkEntityId: &ipOcid,
+		})
+
+		err = cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	return nil
+}
+
+// DeleteRoute delete a route entry, but OCI only has Update of the entire RouteTable
+func (cp *CloudProvider) DeleteRoute(ctx context.Context, clusterName string, route *cloudprovider.Route) error {
+	glog.V(6).Infof("Delete Route: %v", route)
+
+	for routeID := range routeTableIds {
+		oRoute, err := cp.client.Networking().GetRouteTable(ctx, routeID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		dirty := false
+
+		for i, r := range oRoute.RouteRules {
+			if *r.CidrBlock != route.DestinationCIDR {
+				continue
+			}
+
+			dirty = true
+
+			glog.V(6).Infof("Deleting Route: %s %s", route.DestinationCIDR, route.TargetNode)
+			oRoute.RouteRules = append(oRoute.RouteRules[:i], oRoute.RouteRules[i+1:]...)
+		}
+
+		if dirty {
+			glog.V(6).Infof("Updating Route: %s %v", routeID, oRoute.RouteRules)
+			err := cp.client.Networking().UpdateRouteTable(ctx, routeID, oRoute.RouteRules)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// subnet -> routetable
+// subnets cannot change their route tables, so it's safe to cache
+var subnetIds map[string]string
+
+// routetable -> exists
+// we may see new route tables if we see new subnets
+var routeTableIds map[string]bool
+
+// ListRoutes for OCI, discover all instances and their subnets and those route tables to iterate over
+func (cp *CloudProvider) ListRoutes(ctx context.Context, clusterName string) ([]*cloudprovider.Route, error) {
+	glog.V(6).Info("Listing Routes: ", clusterName)
+
+	if subnetIds == nil {
+		subnetIds = make(map[string]string)
+	}
+
+	newRouteTableIds := make(map[string]bool)
+
+	nodeList, err := cp.NodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// Get subnets and their route tables. Mark which route tables are in use.
+	ipToNodeLookup := make(map[string]*v1.Node)
+	for _, node := range nodeList {
+		err := cp.mapNodeSubnetToRouteTable(ctx, node, newRouteTableIds)
+		if err != nil {
+			return nil, err
+		}
+		ip := util.NodeInternalIP(node)
+		ipToNodeLookup[ip] = node
+	}
+
+	routeTableIds = newRouteTableIds
+
+	// ultimately we need to find all the routes from all route tables
+	// that are germane to this cluster, so if the destination cidr isn't
+	// in the cluster range skip, also if we find mismatch destinations set
+	// target to empty to invalidate and force a refresh
+	var destinationsMap = make(map[string]string)
+
+	for id := range routeTableIds {
+		glog.V(6).Infof("Listing routes for id %s", id)
+
+		oRoute, err := cp.client.Networking().GetRouteTable(ctx, id)
+		if err != nil {
+			glog.V(6).Infof("Failed to get routes for id %s", id)
+			return nil, errors.WithStack(err)
+		}
+
+		glog.V(6).Infof("Route entries: %v", oRoute.RouteRules)
```

Comment:
Route rules potentially have 100+ things right? Not sure i'd log this even during debug.